### ISSUE TITLE
What is the "inflight request" mean?

### ISSUE THREAD

**Issue by monsterlyg** [triton-inference-server/server#6355]
I try to use "httpclient.InferenceServerClient.async_infer" function to issue inference. The "async_infer" is decribed as 
```
Run asynchronous inference using the supplied 'inputs' requesting
        the outputs specified by 'outputs'. Even though this call is
        non-blocking, however, the actual number of concurrent requests to
        the server will be limited by the 'concurrency' parameter specified
        while creating this client. In other words, if the ***inflight***
        async_infer exceeds the specified 'concurrency', the delivery of
        the exceeding request(s) to server will be blocked till the slot is
        made available by retrieving the results of previously issued requests.
```
What the "inflight" state means ? And another strange phenomenon is that I put InferAsyncRequest returned by async_infer function into a queue.Queue  and get_result in another thread like below. The delay time grows over time. In my understanding, the server will reject the request if the max_queue_size is full.  So even the client send the requests whose num is bigger than the server can handle. The async_request can get result(even null). But the fact is the queue(processed_reqs) size also grows. This means there exists requests not been processed. I am confused with this and no solution found after many tries. Hope for some explanation.
```
while True:
    time1 = time.time()
    async_req = safe_client.triton_client.async_infer(model_name=FLAGS.model_name,
                                                    inputs=inputs, outputs=outputs)
     processed_reqs.put((time1,async_req))
    time.sleep(0.1)
...
while True:
    time1, async_request = processed_reqs.get(timeout=FLAGS.queue_timeout)
    result = async_request.get_result()
    time2 = time.time()
    delay  = time2 - time1
    ...
```

**Comment by monsterlyg** [comment]
By the way, once the async_infer is called, does mean that the request is to be sent to server ?

**Comment by nnshah1** [comment]
`Inflight` in this context refers to all requests that have been issued via `async_infer`.
`concurrency` in this context refers to the number of parallel requests the http client will send to the server.

For example, if the concurrency is 5, and you've called `async_infer` 10 times with no delay:

1) the client will send 5 requests to the server in parallel. These would be available for batching,  evaluated against max_queue_size etc. and get responses from the server as soon as available.
2) the client will hold the next 5 requests on the client side - until one of the original 5 requests return. 

Thus `concurrency` limits the rate with which the client sends requests to the server. It is purely a client side setting - and does directly impact the server behavior.

**Comment by monsterlyg** [comment]
> `Inflight` in this context refers to all requests that have been issued via `async_infer`. `concurrency` in this context refers to the number of parallel requests the http client will send to the server.
> 
> For example, if the concurrency is 5, and you've called `async_infer` 10 times with no delay:
> 
> 1. the client will send 5 requests to the server in parallel. These would be available for batching,  evaluated against max_queue_size etc. and get responses from the server as soon as available.
> 2. the client will hold the next 5 requests on the client side - until one of the original 5 requests return.
> 
> Thus `concurrency` limits the rate with which the client sends requests to the server. It is purely a client side setting - and does directly impact the server behavior.

Suppose the concurrency is 10, and the client call 10 times. But the server can only handle one request at a time and the max_queue_size is 1. What will happen?

**Comment by monsterlyg** [comment]
@nnshah1  No. Http async_infer returns InferAsyncRequest and raises InferenceServerException. In that case, InferenceServerException error is expected. Besides, Grpc async_infer seems not call true asychronous request ? The call function ends with a delay in which the server has the request processed. It seems that the 'callback' is blocked.

**Comment by monsterlyg** [comment]
```
def callback_res(timestamp, processed_requests, result, error):
    cost_t = time.time() - timestamp
    if error:
        processed_requests.put((cost_t, error))
    else:
        processed_requests.put((cost_t, result))
processed_reqs = queue.Queue()
start_timestamp = time.time().

 tritonclient.grpc.async_infer(model_name=FLAGS.model_name,
                                                        inputs=inputs,
                                                        callback=partial(callback_res, 
                                                                         start_timestamp, processed_reqs),
                                                        outputs=outputs)
start_t2 = time.time()
logger.debug(f"cost time : {(start_t2-start_timestamp):.4f} ")
```
In my case, ((start_t2-start_timestamp) is about 0.3s, which is almost equal to the time the srever cost. So it seems the same as 'infer' which is blocked.

**Comment by monsterlyg** [comment]
> ```
> def callback_res(timestamp, processed_requests, result, error):
>     cost_t = time.time() - timestamp
>     if error:
>         processed_requests.put((cost_t, error))
>     else:
>         processed_requests.put((cost_t, result))
> processed_reqs = queue.Queue()
> start_timestamp = time.time().
> 
>  tritonclient.grpc.async_infer(model_name=FLAGS.model_name,
>                                                         inputs=inputs,
>                                                         callback=partial(callback_res, 
>                                                                          start_timestamp, processed_reqs),
>                                                         outputs=outputs)
> start_t2 = time.time()
> logger.debug(f"cost time : {(start_t2-start_timestamp):.4f} ")
> ```
> 
> In my case, ((start_t2-start_timestamp) is about 0.3s, which is almost equal to the time the srever cost. So it seems the same as 'infer' which is blocked. @nnshah1
